planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite
+    count(*) -> count(not null column)
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(constant)
+
+Pattern:
+LogcialAggregation: count(constant)
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns from datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the narrowest not null column from datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns from datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the narrowest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
+
+*/
+
+type countStarRewriter struct {
+}
+
+func (c *countStarRewriter) optimize(ctx context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return c.countStarRewriter(p, opt)
+}
+
+func (c *countStarRewriter) countStarRewriter(p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	// match pattern agg(count(*)) -> datasource
+	if agg, ok := p.(*LogicalAggregation); ok {
+		if len(agg.GroupByItems) == 0 {
+			if dataSource, ok := agg.Children()[0].(*DataSource); ok {
+				for _, aggFunc := range agg.AggFuncs {
+					if aggFunc.Name == "count" && len(aggFunc.Args) == 1 && !aggFunc.HasDistinct {
+						if constExpr, ok := aggFunc.Args[0].(*expression.Constant); ok {
+							if !constExpr.Value.IsNull() {
+								if len(dataSource.Columns) > 0 {
+									rewriteCountConstantToCountColumn(dataSource, aggFunc)
+									continue
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
We can wrap these into the `matchPatternAndRewrite` func and then
```golang
agg, ok := p.(Agg)
if !ok || len(agg.GroupByItems) == 0 {
	return
}
dataSource, ok := agg.Children()[0].(DS)
if !ok {
	return
}
for aggFunc {
	if aggFunc.Name != count || len(aggFunc.Args) > 1 || aggFunc.HasDistinct {
		continue	
	}
	...
}
```

Or just
```golang
if agg, ok := p.(Agg); ok && len(agg.GroupByItems) > 0 {
	if check datasource {
		for aggFunc {
			if name == count && len(args) == 1 && !hasDistinct {
				if const, ok := args[0].(const); ok && const is not null && len(dataSource.Columns) >0 {
					rewrite()
				}
			}
		}	
	}
}
```
to fold the multiple for+if code blocks.

Reply:
If I put pattern and rewriter in one function, I can't return early. But I can change it to look better in this way, you can take a look again.